File: src/main/java/com/jozufozu/exnihiloomnia/common/registries/RegistryLoader.java
package com.jozufozu.exnihiloomnia.common.registries;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.jozufozu.exnihiloomnia.ExNihilo;
import com.jozufozu.exnihiloomnia.common.registries.recipes.*;
import net.minecraft.util.JsonUtils;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.event.RegistryEvent;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.registries.IForgeRegistry;
import net.minecraftforge.registries.IForgeRegistryEntry;
import org.apache.commons.io.FilenameUtils;
import org.apache.commons.io.IOUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.*;
import java.net.URISyntaxException;
import java.net.URL;
import java.nio.file.Files;
import java.util.MissingResourceException;
import java.util.Stack;
import java.util.function.Consumer;
import java.util.function.Function;

@Mod.EventBusSubscriber(modid = ExNihilo.MODID)
public class RegistryLoader
{
    public static final boolean DEV_MODE = true; //Reloads all registries from assets every time
    
    public static Gson GSON = (new GsonBuilder()).setPrettyPrinting().disableHtmlEscaping().create();
    private static final Logger LOGGER = LogManager.getLogger("Ex Nihilo");
    
    private static final Stack<String> CONTEXT = new Stack<>();
    
    @SubscribeEvent
    public static void loadCompost(RegistryEvent.Register<CompostRecipe> event)
    {
        genericLoad(event, "/registries/composting", CompostRecipe::deserialize);
    }
    
    @SubscribeEvent
    public static void loadFluidCrafting(RegistryEvent.Register<FluidCraftingRecipe> event)
    {
        genericLoad(event, "/registries/fluidcrafting", FluidCraftingRecipe::deserialize);
    }
    
    @SubscribeEvent
    public static void loadFluidMixing(RegistryEvent.Register<FluidMixingRecipe> event)
    {
        genericLoad(event, "/registries/fluidmixing", FluidMixingRecipe::deserialize);
    }
    
    @SubscribeEvent
    public static void loadFermenting(RegistryEvent.Register<FermentingRecipe> event)
    {
        genericLoad(event, "/registries/fermenting", FermentingRecipe::deserialize);
    }
    
    @SubscribeEvent
    public static void loadSieve(RegistryEvent.Register<SieveRecipe> event)
    {
        genericLoad(event, "/registries/sieve", SieveRecipe::deserialize);
    }
    
    @SubscribeEvent
    public static void loadHammer(RegistryEvent.Register<HammerRecipe> event)
    {
        genericLoad(event, "/registries/hammering", HammerRecipe::deserialize);
    }
    
    @SubscribeEvent
    public static void loadMelting(RegistryEvent.Register<MeltingRecipe> event)
    {
        genericLoad(event, "/registries/melting", MeltingRecipe::deserialize);
    }
    
    @SubscribeEvent
    public static void loadHeat(RegistryEvent.Register<HeatSource> event)
    {
        genericLoad(event, "/registries/heat", HeatSource::deserialize);
    }
    
    private static <T extends IForgeRegistryEntry<T>> void genericLoad(RegistryEvent.Register<T> event, String resourcePath, Function<JsonObject, T> deserializer)
    {
        IForgeRegistry<T> registry = event.getRegistry();
        
        pushCtx("Registry: " + event.getName().toString());
        
        copyIfUnconfigured(resourcePath);
        
        try
        {
            File root = new File(ExNihilo.PATH, resourcePath);
    
            File[] files = root.listFiles((dir, name) -> "json".equals(FilenameUtils.getExtension(name)));
            
            if (files == null)
            {
                return;
            }
            
            for (File file : files)
            {
                String fileName = FilenameUtils.getName(FilenameUtils.removeExtension(file.getName())).replaceAll("\\\\", "/");
                ResourceLocation resourceLocation = new ResourceLocation(ExNihilo.MODID, fileName);
                BufferedReader reader = null;
    
                try
                {
                    try
                    {
                        reader = new BufferedReader(new FileReader(file));
            
                        pushCtx("Entry: " + resourceLocation.toString());
            
                        T registryEntry = deserializer.apply(JsonUtils.fromJson(GSON, reader, JsonObject.class));
            
                        registryEntry.setRegistryName(resourceLocation);
            
                        registry.register(registryEntry);
                    }
                    catch (JsonParseException jsonparseexception)
                    {
                        error("Parsing error, " + jsonparseexception.getMessage());
                    }
                    catch (IOException ioexception)
                    {
                        error("Couldn't read " + event.getName() + " entries " + resourceLocation + " from " + file.getPath(), ioexception);
                    }
                }
                finally
                {
                    IOUtils.closeQuietly(reader);
                    popCtx();
                }
            }
        }
        catch (Exception ioException)
        {
            error("Couldn't get a list of all " + event.getName() + " registry files", ioException);
        }
        finally
        {
            clearCtx();
        }
    }
    
    public static void loadSingleJson(String resourcePath, Consumer<JsonObject> toDo)
    {
        File toLoad = new File(ExNihilo.PATH, resourcePath);
    
        try
        {
            BufferedReader reader = new BufferedReader(new FileReader(toLoad));
    
            toDo.accept(JsonUtils.fromJson(GSON, reader, JsonObject.class));
        }
        catch (FileNotFoundException e)
        {
            LOGGER.error("Could not find resource: '" + resourcePath + "'");
        }
    }
    
    public static void copySingle(String resourcePath)
    {
        File to = new File(ExNihilo.PATH, resourcePath);
    
        if (DEV_MODE)
        {
            recursiveDelete(to);
        }
        
        if (to.exists())
        {
            return;
        }
        
        InputStream in = null;
        
        try
        {
            URL resource = RegistryLoader.class.getResource("/assets/exnihiloomnia" + resourcePath);
            
            if (resource == null)
            {
                return;
            }
            
            File file = new File(resource.toURI());
            
            in = new FileInputStream(file);
            Files.copy(in, to.toPath());
        }
        catch (IOException | URISyntaxException e)
        {
        
        }
        finally
        {
            IOUtils.closeQuietly(in);
        }
    }
    
    public static void copyIfUnconfigured(String resourcePath)
    {
        File config = new File(ExNihilo.PATH, resourcePath);
        
        if (DEV_MODE)
        {
            recursiveDelete(config);
        }
        
        if (config.exists())
            return;
        
        config.mkdirs();
        
        try
        {
            File registryPath = new File(RegistryLoader.class.getResource("/assets/exnihiloomnia" + resourcePath).toURI());
    
            File[] files = registryPath.listFiles((dir, name) -> "json".equals(FilenameUtils.getExtension(name)));
            
            if (files == null)
            {
                throw new MissingResourceException("No files to load!", "RegistryLoader", "/assets/exnihiloomnia" + resourcePath);
            }
    
            for (File file : files)
            {
                File to = new File(config, FilenameUtils.getName(file.getName()));
                InputStream in = null;
    
                try
                {
                    in = new FileInputStream(file);
                    Files.copy(in, to.toPath());
                }
                finally
                {
                    IOUtils.closeQuietly(in);
                }
            }
        }
        catch (IOException | URISyntaxException uriSyntaxException)
        {
            error("Error copying assets to config: " + uriSyntaxException.getMessage());
        }
    }
    
    private static void recursiveDelete(File file)
    {
        if (!file.exists() || !file.getAbsolutePath().contains("config"))
            return;
        
        if (file.isDirectory())
        {
            File[] files = file.listFiles();
            
            if (files == null)
                return;
            
            for (File file1 : files)
            {
                recursiveDelete(file1);
            }
            
            file.delete();
        }
        else
        {
            file.delete();
        }
    }
    
    public static String pushCtx(String ctx)
    {
        return CONTEXT.push(ctx);
    }
    
    public static String popCtx()
    {
        return CONTEXT.pop();
    }
    
    public static void clearCtx()
    {
        CONTEXT.clear();
    }
    
    public static void error(String s)
    {
        StringBuilder out = new StringBuilder();
    
        for (String s1 : CONTEXT)
        {
            out.append("[").append(s1).append("]");
        }
        
        out.append(" ").append(s);
        
        LOGGER.error(out.toString());
    }
    
    public static void error(String s, Throwable t)
    {
        StringBuilder out = new StringBuilder();
        
        for (String s1 : CONTEXT)
        {
            out.append("[").append(s1).append("]");
        }
        
        out.append(" ").append(s);
        
        LOGGER.error(out.toString(), t);
    }
    
    public static void error(Throwable t)
    {
        StringBuilder out = new StringBuilder();
        
        for (String s1 : CONTEXT)
        {
            out.append("[").append(s1).append("]");
        }
        
        LOGGER.error(out.toString(), t);
    }
}

File: src/main/kotlin/com/jozufozu/exnihiloomnia/common/registries/RegistryLoader.kt
package com.jozufozu.exnihiloomnia.common.registries

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonObject
import com.google.gson.JsonParseException
import com.jozufozu.exnihiloomnia.ExNihilo
import com.jozufozu.exnihiloomnia.common.ModConfig
import net.minecraft.util.JsonUtils
import net.minecraft.util.ResourceLocation
import net.minecraftforge.common.crafting.JsonContext
import net.minecraftforge.registries.IForgeRegistryEntry
import org.apache.commons.io.FilenameUtils
import org.apache.commons.io.IOUtils
import org.apache.logging.log4j.LogManager
import java.io.*
import java.net.URISyntaxException
import java.nio.file.*
import java.util.*
import kotlin.streams.asSequence


object RegistryLoader {
    val GSON: Gson = GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create()
    val CONTEXT = JsonContext(ExNihilo.MODID)

    private val LOGGER = LogManager.getLogger("Ex Nihilo")

    private val loggerLayers = Stack<String>()

    fun loadOres() {
        val timeThen = System.nanoTime()
        RegistryManager.ORES.load()
        val timeNow = System.nanoTime()

        clearCtx()
        LOGGER.info("Finished loading ores, took ${(timeNow - timeThen).toFloat() * 1e-9} seconds")
    }

    fun loadRecipes() {
        val timeThen = System.nanoTime()
        RegistryManager.COMPOST.load()
        RegistryManager.FLUID_CRAFTING.load()
        RegistryManager.FLUID_MIXING.load()
        RegistryManager.LEAKING.load()
        RegistryManager.FERMENTING.load()
        RegistryManager.SUMMONING.load()
        RegistryManager.SIFTING.load()
        RegistryManager.HAMMERING.load()
        RegistryManager.MELTING.load()
        RegistryManager.HEAT.load()
        RegistryManager.MESH.load()
        val timeNow = System.nanoTime()

        clearCtx()
        LOGGER.info("Finished loading all registries, took ${(timeNow - timeThen).toFloat() * 1e-9} seconds")
    }

    fun <T : IForgeRegistryEntry<T>> genericLoad(registry: ReloadableRegistry<T>, resourcePath: String, deserializer: (JsonObject) -> T?) {
        val clear = pushCtx(registry.registryName.resourcePath)

        try {
            getFilesToLoad(resourcePath) { stream, name ->
                restoreCtx(clear)
                val resourceLocation = ResourceLocation(ExNihilo.MODID, FilenameUtils.getName(FilenameUtils.removeExtension(name)).replace("\\\\".toRegex(), "/"))
                try {
                    pushCtx(resourceLocation.resourcePath)

                    JsonUtils.fromJson(GSON, stream, JsonObject::class.java)?.run(deserializer)?.let {
                        it.registryName = resourceLocation
                        registry.register(it)
                    }
                } catch (jsonparseexception: JsonParseException) {
                    error("Parsing error, ${jsonparseexception.message}")
                } catch (ioexception: IOException) {
                    error("Couldn't read '${registry.registryName}' entries '$resourceLocation' from $name", ioexception)
                } finally {
                    stream.close()
                }
            }
        } catch (ioException: Exception) {
            error("Couldn't get a list of all '${registry.registryName}' registry files", ioException)
        } finally {
            clearCtx()
        }
    }

    private fun getFilesToLoad(resourcePath: String, action: (Reader, String) -> Unit) {
        if (ModConfig.enableCraftingCustomization) {
            copyIfUnconfigured(resourcePath)

            val root = File(ExNihilo.PATH, resourcePath)

            (root.listFiles { _, name -> "json" == FilenameUtils.getExtension(name) } ?: return).map { BufferedReader(FileReader(it)) to it.name }.forEach { action(it.first, it.second) }
        } else {
            val path = getFullyQualifiedResourcePath(resourcePath)

            val uri = RegistryLoader::class.java.getResource(path).toURI()

            var fileSystem: FileSystem? = null

            val dir = when (uri.scheme) {
                "file" -> Paths.get(uri)
                "jar" -> {
                    fileSystem = FileSystems.newFileSystem(uri, emptyMap<String, Any>())
                    fileSystem.getPath(path)
                }
                else -> throw Error("Invalid uri scheme loading resources, this in a programmer error!")
            }

            Files.walk(dir).asSequence().filter { it.toString().endsWith(".json") }.map { Files.newBufferedReader(it) to it.toString() }.forEach { action(it.first, it.second) }

            IOUtils.closeQuietly(fileSystem)
        }
    }

    fun loadSingleJson(resourcePath: String, callback: (JsonObject) -> Unit): Boolean {
        val toLoad = if (ModConfig.enableCraftingCustomization) {
            val to = File(ExNihilo.PATH, resourcePath)

            if (!to.exists()) {

                var inputStream: InputStream? = null

                try {
                    val resource = RegistryLoader::class.java.getResource(getFullyQualifiedResourcePath(resourcePath))
                            ?: return false

                    val file = File(resource.toURI())

                    to.mkdirs()
                    inputStream = FileInputStream(file)
                    Files.copy(inputStream, to.toPath(), StandardCopyOption.REPLACE_EXISTING)
                } catch (e: Exception) {
                    error(e)
                    return false
                } finally {
                    IOUtils.closeQuietly(inputStream)
                }
            }

            to
        } else {
            val resource = RegistryLoader::class.java.getResource(getFullyQualifiedResourcePath(resourcePath))
                    ?: return false

            File(resource.toURI())
        }

        try {
            val reader = BufferedReader(FileReader(toLoad))

            val json = JsonUtils.fromJson(GSON, reader, JsonObject::class.java)
            if (json == null) {
                LOGGER.error("Could not load JSON file '$resourcePath'")
                return false
            }
            callback(json)
            return true
        } catch (e: FileNotFoundException) {
            LOGGER.error("Could not find resource: '$resourcePath'")
        }

        return false
    }

    private fun getFullyQualifiedResourcePath(path: String) = "/data/exnihiloomnia/exnihiloomnia$path"

    fun copyIfUnconfigured(resourcePath: String) {
        val config = File(ExNihilo.PATH, resourcePath)

        if (config.exists())
            return

        config.mkdirs()

        try {
            val registryPath = File(RegistryLoader::class.java.getResource(getFullyQualifiedResourcePath(resourcePath)).toURI())

            val files = registryPath.listFiles { _, name -> "json" == FilenameUtils.getExtension(name) }
                    ?: throw MissingResourceException("No files to load!", "RegistryLoader", getFullyQualifiedResourcePath(resourcePath))

            for (file in files) {
                val to = File(config, FilenameUtils.getName(file.name))
                var inputStream: InputStream? = null

                try {
                    inputStream = FileInputStream(file)
                    Files.copy(inputStream, to.toPath())
                } finally {
                    IOUtils.closeQuietly(inputStream)
                }
            }
        } catch (uriSyntaxException: IOException) {
            error("Error copying assets to config: ${uriSyntaxException.message}")
        } catch (uriSyntaxException: URISyntaxException) {
            error("Error copying assets to config: ${uriSyntaxException.message}")
        }
    }

    fun pushCtx(ctx: String): Int {
        loggerLayers.push(ctx)
        return loggerLayers.size
    }

    fun pushPopCtx(ctx: String): Int {
        popCtx()
        return pushCtx(ctx)
    }

    fun restoreCtx(size: Int) {
        while (loggerLayers.size > size) loggerLayers.pop()
    }

    fun popCtx(): String {
        return loggerLayers.pop()
    }

    fun clearCtx() {
        loggerLayers.clear()
    }

    fun error(s: String?) {
        val out = StringBuilder()

        for (s1 in loggerLayers) {
            out.append("[").append(s1).append("]")
        }

        out.append(" ").append(s)

        LOGGER.error(out.toString())
    }

    fun warn(s: String?) {
        val out = StringBuilder()

        for (s1 in loggerLayers) {
            out.append("[").append(s1).append("]")
        }

        out.append(" ").append(s)

        LOGGER.warn(out.toString())
    }

    fun error(s: String, t: Throwable) {
        val out = StringBuilder()

        for (s1 in loggerLayers) {
            out.append("[").append(s1).append("]")
        }

        out.append(" ").append(s)

        LOGGER.error(out.toString(), t)
    }

    fun error(t: Throwable) {
        val out = StringBuilder()

        for (s1 in loggerLayers) {
            out.append("[").append(s1).append("]")
        }

        LOGGER.error(out.toString(), t)
    }
}
